File: todo-api-rs/src/application/use_cases/todo/list.rs
use crate::application::dtos::todo::list::{
    ListTodosError, ListTodosInput, ListTodosOutput, TodosList,
};
use crate::application::repositories::todo::{ListError, ListQuery, TodoRepository};
use crate::domain::use_case::UseCase;

#[derive(Debug)]
pub struct ListTodosUseCase<T> {
    repository: T,
}

impl<T: TodoRepository> ListTodosUseCase<T> {
    pub fn new(repository: T) -> Self {
        Self { repository }
    }
}

impl<T: TodoRepository> UseCase<ListTodosInput, ListTodosOutput> for ListTodosUseCase<T> {
    async fn exec(self, input: ListTodosInput) -> ListTodosOutput {
        let query = ListQuery {
            page: input.page,
            per_page: input.per_page,
            title: input.title,
        };

        let list = self.repository.list(query).await.map_err(|err| match err {
            ListError::Internal(err) => ListTodosError::Internal(err),
        })?;

        Ok(TodosList {
            count: list.count,
            items: list.items,
            page: input.page,
            per_page: input.per_page,
        })
    }
}

File: todo-api-js/src/application/use-cases/todo/list.ts
import { Err } from "@domain/utils/result";
import { InternalError } from "@domain/utils/exception";
import { type UseCase } from "@domain/use-case";
import { type TodoRepository } from "@application/repositories/todo";
import { type ListTodosInput, type ListTodosOutput } from "@application/dtos/todo/list";

export class ListTodosUseCase implements UseCase<ListTodosInput, ListTodosOutput> {
    constructor(private repository: TodoRepository) {}

    async exec(input: ListTodosInput): Promise<ListTodosOutput> {
        try {
            return this.repository.listPaginated(input);
        } catch (e) {
            const error = new InternalError(e instanceof Error ? e : undefined);
            return new Err(error);
        }
    }
}
